### PR TITLE
#5856: added option to create vertically-oriented stem plots

### PR DESCRIPTION
**Pull request by vecuenca**
This is an attempt at implementing #5856.

I introduced a 'vertical' parameter  that, upon being passed as true to the stem method, creates a vertical stem plot.

### CODE REVIEW

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -2367,11 +2367,14 @@ def stem(self, *args, **kwargs):
...
           stem(x, y, linefmt='b-', markerfmt='bo', basefmt='r-')
 
-        A stem plot plots vertical lines (using *linefmt*) at each *x*
+        A stem plot, by default, plots vertical lines (using *linefmt*) at each *x*
         location from the baseline to *y*, and places a marker there
         using *markerfmt*.  A horizontal line at 0 is is plotted using
         *basefmt*.
 
+        It is possible to create a vertically-oriented stem plot
+        by providing vertical=True.
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: Can you convert this to numpydoc format and enumerate all of the things that are popped out of the kwargs in the docstring?

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -2420,22 +2423,38 @@ def stem(self, *args, **kwargs):
         except IndexError:
             basefmt = kwargs.pop('basefmt', 'r-')
 
+        # Check if the user wants a vertical stem plot
+        vertical = kwargs.pop('vertical', None)
+        if vertical is None:
+            vertical = False
+
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: It is a bigger change, but could you refactor this to have on `if` block up here which does the flipping (and maybe rename x, y?) and then the rest of the logic is the same.

File: lib/matplotlib/axes/_axes.py [hunk 3]
```diff
@@ -2420,22 +2423,38 @@ def stem(self, *args, **kwargs):
         except IndexError:
             basefmt = kwargs.pop('basefmt', 'r-')
 
+        # Check if the user wants a vertical stem plot
+        vertical = kwargs.pop('vertical', None)
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: might as well have pop default to `False`

This is one function that would greatly benefit from dropping python 2 as we could write the signature as

```python
def stem(*args, vertical=False, bottom=None, label=None, **kwargs):
```

**Comment by vincent-cuenca** [comment]
Is this what you were asking for regarding the flipping?

**Comment by tacaswell** [comment]
Yes, but I think you can do the same with the other two `if vertical` blocks.

File: lib/matplotlib/axes/_axes.py [hunk 4]
```diff
@@ -2420,9 +2444,16 @@ def stem(self, *args, **kwargs):
         except IndexError:
             basefmt = kwargs.pop('basefmt', 'r-')
 
+        # Check the orientation variable to see if the user
+        # wants a vertical or horizontal stem plot
+        orientation = kwargs.pop('orientation', 'horizontal')
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: should probably do some validation here to make sure 'aardvark' or the like does not get passed in.

File: lib/matplotlib/axes/_axes.py [hunk 5]
```diff
@@ -2430,12 +2461,22 @@ def stem(self, *args, **kwargs):
 
         stemlines = []
         for thisx, thisy in zip(x, y):
-            l, = self.plot([thisx, thisx], [bottom, thisy], linefmt,
-                           label="_nolegend_")
+            if orientation == 'vertical':
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: Can this if block be eliminated?

**Comment by tacaswell** [comment]
Is this the PR that will be the combined effort?

This will also need a test + a whats_new entry.

**Comment by vincent-cuenca** [comment]
Yup, this is the combined PR.
Re: testing, would test_axes.py be an appropriate place to make one?
Im not too sure where to put the whats_new entry as well.

**Comment by tacaswell** [comment]
Where ever the existing tests of stemplot live is the correct place to put these.

See https://github.com/matplotlib/matplotlib/tree/master/doc/users/whats_new for the whats_new entry instructions.

**Comment by yongzhez** [comment]
Hi I'm working with @vincent-cuenca and I was wondering what the progress on this pull request is. What else is needed to complete this pull request?

File: lib/matplotlib/axes/_axes.py [hunk 6]
```diff
@@ -2374,17 +2376,39 @@ def stem(self, *args, **kwargs):
 
         If no *x* values are provided, the default is (0, 1, ..., len(y) - 1)
 
-        Return value is a tuple (*markerline*, *stemlines*,
-        *baseline*).
+        Optional parameters:
```
Comment (jenshnielsen) [inline lib/matplotlib/axes/_axes.py]: This does not conform with numpy doc specs see https://github.com/matplotlib/matplotlib/pull/6240 for a similar fix. 
If you change it to `Other Parameters` it should work. As it is this section will not be in the web generated version of the docs. 

See https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt for more numpy docs info. 

The output from the docs build should this is in https://travis-ci.org/matplotlib/matplotlib/jobs/118347853#L1036

**Comment by jenshnielsen** [comment]
Apart from inline comment 👍  from me.

**Comment by jklymak** [comment]
This actually seemed done, but never got a final review and now needs a rebase.  It'd be a reasonable good first issue for someone to revive it.

**Comment by QuLogic** [comment]
This PR is too old to do a direct push, so I've opened a new one, noted above.

**Comment by timhoffm** [comment]
Closing in favor of #18187.

@vecuenca thanks for the contribution. Your name is still present in the commits of #18187, so you will still get credit once it is merged.

**Comment by QuLogic** [comment]
Note, this _should_ have closed automatically.

**Pull request closed without merging**